sklearn_numba_dpex/examples/kmeans_numba_dpex.py

```diff
@@ -0,0 +1,628 @@
+#             row = i // dim
+#             if row > group_nb_clusters:
+#                 break
+#             col = i % dim
+#             local_centroids[row, col] = float32(0.)
+
+#         dpex.barrier()
+
+#         current_pt = thread_starting_pt
+#         current_dim = thread_starting_dim
+#         current_cluster_idx = local_assignments_idx[current_pt]
+#         for pi in range(p):
+#             if current_pt >= n:
+#                 break
+#             e = x[current_pt, current_dim]
+#             dpex.atomic.add(local_centroids, (current_cluster_idx, current_dim), e)
+#             if current_dim == 0:
+#                 dpex.atomic.add(partial_cluster_count, current_cluster_idx, 1)
+#             current_dim = (current_dim + 1) % dim
+#             if current_dim == 0:
+#                 current_pt += 1
+#                 current_cluster_idx = local_assignments_idx[current_pt]
+
+#         dpex.barrier()
+
+#         current_row = starting_entry // dim
+#         current_cluster_idx = local_cluster_to_global[row]
+#         for i in range(starting_entry, starting_entry + nb_centroid_entries_per_thread):
+#             row = i // dim
+#             if row > group_nb_clusters:
+#                 break
+#             if row != current_row:
+#                 current_cluster_idx = local_cluster_to_global[row]
+#                 current_row = row
+#             col = i % dim
+#             if col == 0:
+#                 dpex.atomic.add(centroid_counts, current_cluster_idx, partial_cluster_count[row])
+#             dpex.atomic.add(centroids, (current_cluster_idx, dim), local_centroids[row, col])
+
+
+#         current_row = starting_entry // dim
+#         current_cluster_idx = local_cluster_to_global[row]
+#         for i in range(starting_entry, starting_entry + nb_centroid_entries_per_thread):
+#             row = i // dim
+#             if row > group_nb_clusters:
+#                 break
+#             if row != current_row:
+#                 current_cluster_idx = local_cluster_to_global[row]
+#                 current_row = row
+#             col = i % dim
+#             if col == 0:
+#                 dpex.atomic.add(centroid_counts, current_cluster_idx, partial_cluster_count[row])
+#             dpex.atomic.add(centroids, (current_cluster_idx, dim), local_centroids[row, col])
+
+
+def kmeans_fit_numba_dpex(x_train,
+                          y_train,
+                          n_clusters=128,
+                          init="random",
+                          n_init=10,
+                          max_iter=300,
+                          tol=1e-4,
+                          random_state=123,
+                          algorithm="full",
+                          copy_x=False):
+
+    #!!!: x_train, y_train must be in row_major order (C-style)
+    # and should be float32
+
+    # TODO: write a kernel for this ?
+    # x_train[:] = x_train - x_train.mean(axis=0)
```

Comment:
I think numba already compiles numpy operations and in this case, I would suspect that numba-dpex also does.

Reply:
numba_dpex is more like numba.cuda rather than numba, you have a very limited pool of tools within those namespaces